File: 148/sort-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void print(struct ListNode* head){
    while(head!=NULL){
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");
}
struct ListNode* sortList(struct ListNode* head){
    //print(head);
    //只要能找到一半,便能將數字 divide & conquer 做出 merge sort
    if(head==NULL) return NULL;
    if(head->next==NULL) return head;

    struct ListNode* fast = head;
    struct ListNode* slow = head;
    while(fast->next!=NULL && fast->next->next!=NULL){
    //while(fast!=NULL && fast->next!=NULL){不能用這行,不然會切不平均
        fast = fast->next->next;
        slow = slow->next;
    }
    
    struct ListNode* list2 = slow->next;
    slow->next = NULL;

    struct ListNode * sorted1 = sortList(head);
    struct ListNode * sorted2 = sortList(list2);

    struct ListNode ans;
    struct ListNode * next = &ans;
    while(sorted1!=NULL && sorted2!=NULL){
        if(sorted1->val <= sorted2->val){
            next->next = sorted1;
            next = next->next;
            sorted1 = sorted1->next;
        }else{
            next->next = sorted2;
            next = next->next;
            sorted2 = sorted2->next;
        }
    }
    if(sorted1==NULL) next->next = sorted2;
    else next->next = sorted1;
    
    return ans.next;
}

File: 148/sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 使用 merge sort 來解決此題，先搞定終止條件
        if head==None or head.next ==None: return head

        # 接下來利用 fast,slow 來找到中間點
        fast, slow, prev = head, head, head
        while fast!=None and fast.next!=None:
            fast = fast.next.next
            prev = slow # 備份，以便等一下要斷開
            slow = slow.next
        # 離開迴圈時, slow是中間點
        prev.next = None # 斷開，讓head只有前半斷
        list1 = self.sortList(head)
        list2 = self.sortList(slow)

        # Merge
        ans = ListNode()
        next = ans
        while list1!=None and list2!=None:
            if list1.val < list2.val:
                next.next = list1
                list1 = list1.next
                next = next.next
            else:
                next.next = list2
                list2 = list2.next
                next = next.next
        if list1==None:
            next.next = list2
        else:
            next.next = list1
        return ans.next
